### ISSUE TITLE
[css-page-3] Add orientation descriptor

### ISSUE THREAD

**Issue by chrishtr** [w3c/csswg-drafts#4491]
**TL;DR**
Add the ability to rotate the rendered output of a page, for use in cases such as PDF output, and as a hint to printing code & devices.

**Details**
Currently, the level 3 paged media spec has a [size](https://drafts.csswg.org/css-page-3/#page-size-prop) property. This specifies the per-page constraints on layout.

It would be useful to also have an `orientation` descriptor, which could have values of `portrait` and `landscape` (*). This indicates the orientation of the final rendered page itself, without affecting layout. For example, if content had:

```
size: letter landscape
orientation: portrait
```

It would mean to layout the content in landscape layout mode, but then rotate the resulting rendered page sideways. If the print ended up in a PDF, then a user viewing the PDF would see the page visually rendered with height taller than width, and the content rotated 90 degrees.

This is useful in order to customize the visual display of a PDF, which is a very common way to "print" web pages.

(*) Alternatively, it could have three (four?) values: none, rotate-right, rotate-left (, upside-down?)

@xfq @fantasai

**Comment by heycam** [comment]
No strong opinion on the proposal here.  This would map directly to the `Rotate` entry on the page object in the resulting PDF I guess?  (I looked up Prince's documentation expecting there to be some `-prince`-prefixed way to control this but I couldn't find anything.)

**Comment by faceless2** [comment]
Printing devices will all rotate the pages to fit the available media anyway, they won't need the hint. Yes, if we were implementing this we'd map to the Rotate entry on the PDF Page object, which can take any multiple of 90 as a valid value.

Presumably the intention here is to try and get all the pages the same size on screen? So if you have a single landscape page halfway through a document, for example, you can rotate it to make it visually the same dimensions as the rest of the document? I can see an argument for that - it would help PDF viewers to choose the initial display settings fo the document. But I can't think of any other reason for this myself.

If that's the case then I personally prefer "orientation: portrait", which could leave pages that are originally portrait pages untouched and rotate only landscape pages, over "orientation: 90" or "orientation: rotate-right"

**Comment by AmeliaBR** [comment]
An example of the case of rotated _content_ within portrait page layout. Again, I think this is handled by writing-mode on the rotated table:

![Screenshot of two PDF pages: left page has regular horizontal text, right page has a large table rotated. Footers and margins are the same in both.](https://user-images.githubusercontent.com/9876129/70647086-6c8fd080-1c05-11ea-8a9d-fb0db3d2f8ea.png)

**Comment by AmeliaBR** [comment]
In contrast, the use case being discussed here is where the complete layout of certain pages (headers/footers, and any margins defined for the page) uses a different orientation from the rest, but you want to display all the pages oriented as they would come out of the printer. As Tab says, this would also ensure that when printing & binding they are oriented as intended. (And therefore, the property would need to distinguish between rotated clockwise vs counterclockwise!)

But… I couldn't find any good examples of that type of PDF in a quick poke around my computer.

**Comment by AmeliaBR** [comment]
I thought of another use case: imagine the book layout above, with headers/footers in portrait mode, since that looks best for printing. But for viewing the digital file, it would be much more useful if the PDF pages were displayed in landscape mode, so that the table isn't sideways on your screen.

How would that work with these new descriptors?

**Comment by faceless2** [comment]
This sort of thing (lets call this "image 2")?

![twoup](https://user-images.githubusercontent.com/989243/70650818-ada8d480-1c47-11ea-9fa3-7910352c54fb.png)

I had a look. We have thousands of PDF here but I have struggled to find an example of what this issue describes it's trying to achieve. Generally, if you have a landscape page in the middle of a generally portrait document, it's displayed as landscape.

The idea in Amelias third comment, immediately above, is more common: the page has been laid out entirely in portrait, but with some of the content sideways. You then want to rotate that page from portrait to landscape to get it the right way up, like this one (lets call this "image 3"):
![twoup2](https://user-images.githubusercontent.com/989243/70652553-b18a2600-1c4a-11ea-9fff-3987913942fd.png)

**Comment by Crissov** [comment]
I can think of a surprisingly well-known case where the original viewing condition of the medium is _portrait_ (or _spread_), but for special pages, the medium needs to be rotated by a quarter-turn although the viewing mode and page layout remains _portrait_: [**centerfolds**](https://en.wikipedia.org/wiki/Centerfold). In some cases they are not folded at all, but a magazine with A4-spread has an full-sheet A3 poster in the middle. Is that covered by this issue?

**Comment by chrishtr** [comment]
Sorry for the long delay, I got caught up in many other issues.

**TL;DR:**

Introduce a new `orientation` descriptor. It has three values: `default`, `rotate-right`, and `rotate-left`. In the case of PDF output, it is expected to map directly to the `Rotate` PDF command.

Purpose: apply rotation to PDF output in cases where the most important content has a different orientation than the one from layout.

Semantics: the entire content of the page is rotated when `orientation` is applied, including headers and footers and margins. Layout of the page contents, including aspects such as mixed orientation or writing mode, is unaffected.

In cases where page content has mixed orientation, such as headers and footers differing from page contents rotation, this will result in the header being drawn "sideways" in a PDF display, but as mentioned in the use-case analysis below, this is desirable.

---

Detailed comments below.

Some printing folks in Chrome also looked through various data sets and found examples similar to the ones mentioned in previous comments. They also didn't find examples of the entire page rotating with headers, without already drawing the content in a layout PDF page. Instead, it can be common to have content whose internal sizing fits better to landscape than portrait, so as to have *two* rotations in the same page. The reasons include:
1. tabular data that is better presented with those dimensions
2. large images that are wider than they are tall
3. bitmaps generated by an external formatting system, such as an advanced word processor, or external offline or "legacy" system that generates content

In addition, the headers and footers often have orientations that conflict with the orientation of the content, because the printed page is meant to be presented in a book or pamphlet with a single orientation for each page. The reader of the book is expected to turn it when reading the rotated sub-content.

The original motivation for filing this issue was use case 3. In particular, a word processor such as Google Docs has a different layout and formatting system than the web, and for these reasons performs its own layout and may even draw the results to Canvas instead of DOM elements. On top of this, such "web-based" word processors can often support page sizings that are more advanced or specialized than the user agent's. (*)

(*) Examples: Chromium and WebKit only have limited support for variable-size/shape pagination. OTOH, a word processor often wants to be able to print mixed page sizes and orientations. Regardless, however, it's easy to see that there are plenty of offline, legacy or otherwise non- "web-based" word processors or content generation systems that are unable to generate DOM output, but that nevertheless benefit from being integrated into web content for viewing or printing.

**Comment by tabatkins** [comment]
So just to ensure I'm understanding this right:

I have a paged document that's mostly portrait, but I've got a wide table on one page that is best displayed landscape. I set up my `@page` rules appropriately for this, so the table will break onto a `size: landscape;` page on its own.

When viewed as a PDF, every page flows vertically, as normal for English, and the headers and footers on every page agree with the flow of the content - on the portrait pages, `@top` is along the shorter edge; on the landscape pages, `@top` is along the longer edge.  

When printed, because the printer doesn't reorient the sheets of paper internally, they'll all be arranged portrait-style in the printed stack, and the landscape pages (both the content and the margin areas) will be "sideways", with the reader expected to turn the entire book to read those pages.

So this `orientation` descriptor can be used to force the PDF rendering to conform to that of the printed rendering, right? On all my landscape pages, I can set `orientation` to "rotate-left" or "rotate-right" depending on whether the page is `:left` or `:right`, and it'll render exactly as it would in the stack-of-paper.

That's the intended behavior?

**Comment by chrishtr** [comment]
> So this `orientation` descriptor can be used to force the PDF rendering to conform to that of the printed rendering, right? On all my landscape pages, I can set `orientation` to "rotate-left" or "rotate-right" depending on whether the page is `:left` or `:right`, and it'll render exactly as it would in the stack-of-paper.

It's the other way around. As far as the layout engine of the browser would be concerned, each page has a base layout size of portrait, and would be laid out accordingly. Therefore, without an `orientation` descriptor, the tabular pages would look sideways be hard to read. With `orientation`, the PDF output would know to rotate that page to the preferred orientation of the most important (i.e. the tabular) content.

**Comment by tabatkins** [comment]
Okay, so the case is that the thing assembling the pages knows how to produce a page of content that's "sideways" (landscape-oriented content, but using `size: portrait`); the margins are then all portrait-oriented (`@top` along the short edge), regardless of the page content's orientation.

Then, for the purpose of viewing in a browser, you can set `orientation` on the "sideways" pages to flip them into landscape, so they're easier to view as a PDF; the margins are then "sideways" (still along the shorter edge) but that's less important than making the page viewable without rotating your head or your device 90deg.

Okay, so I accept this as being a reasonable solution for your case 3 (legacy system outputting portrait-oriented bitmaps for pages), but it's completely backwards for cases 1 (wide table) and 2 (wide image), which both want to use `size: landscape` on those pages during layout; they just want the margins to be portrait-oriented instead, so that during printing (when all the pages are arranged portrait-oriented) all the margin boxes are consistent.  As currently described, they instead have to use an ordinary `size: portrait` page, but set `writing-mode` to a vertical value, which is a hacky thing to do in horizontal text.  

It also means that we're privileging the "view as printed/bound" case over the "view on the web" cases, requiring the author to know they have to intervene with the `orientation` property to get the page content to display correctly on the web (or PDF). I think we should be doing the opposite, and making web-viewing correct by default (that is, using `size: landscape` on landscape pages), and then have this feature be able to set the margins to the intended orientation.  Tooling (like Docs) trying to use this feature can output either way, so optimizing for them doesn't matter here.

So I think it should instead be pointed as a margin-orienting feature, like `margin-orientation: default | rotate-left | rotate-right`, with `rotate-left` meaning that the margin boxes are all rotated to the left and laid out that way (so @top is along the left side of the page, and laid out like vertical-lr), etc. This would also inform printing software to use the same rotation for the entire page by default, so that the margin boxes are consistent across pages.

Then the "legacy tools output portrait-oriented bitmaps" would use this by putting the intended-landscape image on a `size: landscape` page, hopefully with the image produced landscape to start with (or at worst, using `image-orientation` to rotate the portrait image into landscape), and then `margin-orientation: rotate-*` to indicate which shorter side should be the "top" when printing.

Thoughts?

**Comment by chrishtr** [comment]
> It also means that we're privileging the "view as printed/bound" case over the "view on the web" cases

The orientation descriptor is just for printed output. I think it's ok to have a few CSS properties that are just for controlling printing. Printing is part of the web after all. These orientation descriptors should only be within an @media print section.

For a "view on the web" mode of the same site, if there is content that is best viewed with a landscape aspect ratio, then the site should just lay out the content in a landscape manner; there are many ways to do that already. What is different about printing (or PDF viewing) is that there is a physical or virtual page that is external to the web on which content is printed. The system that might view a PDF can't re-do layout, it can only rotate around the page.

> Then the "legacy tools output portrait-oriented bitmaps" would use this by putting the intended-landscape image on a `size: landscape` page

In addition to my other arguments about why this is not the best way to do it, there is a very practical one, which I alluded to in my previous comment when I said the "limited support". To support `size: landscape` properly without violating specs is very difficult, and a multi-year project for Chromium. I'm presuming the same is true of other UAs to greater or lesser extent.

**Comment by chrishtr** [comment]
> What is different about printing (or PDF viewing) is that there is a physical or virtual page that is external to the web on which content is printed. The system that might view a PDF can't re-do layout, it can only rotate around the page.

Also, in the use cases I mentioned, there is an explicit need for the printed output on paper to differ from an "expected view rotation" when viewed digitally in a PDF viewer.

**Comment by faceless2** [comment]
The only use-case described so far seems to be to "undo" the layout decisions made by some layout engines that rotate landscape content into a portrait format. Have I understood that correctly?

Are there any use-cases for this property beyond that? Without being able to describe a situation where a member of the general public might use this, it seems a little... specialized for a property baked into a public specification.

I also wonder how the following would work:

1. Content with "position: fixed" which appears on every page? Is that rotated or not?
2. Backgrounds promoted from the HTML body or the :root to the page context? Rotated or not?

**Comment by chrishtr** [comment]
@faceless2 

> The only use-case described so far seems to be to "undo" the layout decisions made by some layout engines that rotate landscape content into a portrait format. Have I understood that correctly?

I don't think that's accurate. The use case is to apply a rotation to the "most important" direction of text, in cases where there are multiple orientations on the same page for printed output.

To your questions about deciding whether certain parts of the page should be rotated or not - I think that's not the best way to think about it. It's that the page has a particular printed output that includes rotated text in various combinations, for multiple reasons. `orientation` doesn't say how to rotate various pieces of content in different ways, there are already mechanisms for that. It just says which orientation is the most useful way to view a particular page.

**Comment by faceless2** [comment]
OK. I'm not hugely invested in this one way or another so I will say my piece and go away.

I'm struggling because I can't imagine a general use-case that demonstrates how to get content onto the page that isn't already upright, according to its "most important direction", in the first place.

* extra wide table? Make the page landscape.
* landscape format image? Make the page landscape.

In both cases you would have to work much harder to display this content sideways on a portrait page. For a large image, for example, you'd have to override `image-orientation`, or apply a rotation `transform` to it. Making a large table come out sideways? Well I suppose you could try with `writing-mode` and `transform`.

I can see that there might be some value in rotating the _margin content_ in this case - for example, in Amelia's example above, if I were tasked with laying this table out in CSS I'd make it a landscape page. But if I knew it was being destined for a printed book, I might want to rotate the margin content by 90°, so when it's printed and bound, in portrait, the margin content is the right way up.

I think that's the opposite of your suggested purpose, but I suppose it's literally a question of perspective so perhaps it will do as a use-case.

---

For background: underlying my concerns is that the "Rotate" parameter in PDF is not great. It's not a simple Affine transformation - it rotates the page content, yes, and the coordinates of any annotations on the page. But the internal contents of those annotations are further rotated to try to keep them up the right way up. This doesn't matter if all you're doing is putting out static content, but if you're creating a PDF with form fields on it then creating the page with the correct orientation in the first place is a lot cleaner than creating the page sideways and setting "Rotate" to 90 or (worse) 180.

Clearly your intent is _not_ to replace `size: A4 landscape` with `size: A4; orientation: rotate-right`. That's good, but I am slightly concerned that if this property enters the spec, that's precisely what people will use it for. For that reason, I'd lean towards something like Tab's suggestion of `margin-orientation` myself.

**Comment by chrishtr** [comment]
> * extra wide table? Make the page landscape.
> * landscape format image? Make the page landscape.

If the page is landscape then headers will also be landscape, right? Which is not desired.

**Comment by tabatkins** [comment]
> The orientation descriptor is just for printed output.

As far as I can tell from your use-case description and described behavior, it's the exact opposite. From your description of the use-case, it's for when you're rendering to a PDF or other page-based *screen* media. In the case you described, all the pages are `size: portrait` already, and thus suitable for immediately printing with no additional work; you use `orientation` to get the supposed-to-be-landscape pages to be shown in landscape *on screen*.

My suggested inversion was the "just for printed output" version, where screen/PDF display works correctly by default (by using `size` properly), and then you use `margin-orientation` to rotate how the margins are laid out so they're in a desirable spot when printing.

(@faceless2 hits on this independently, realizing that the semantics they find reasonable are the opposite of what's being presented here.)

> To support size: landscape properly without violating specs is very difficult, and a multi-year project for Chromium.

Okay, *that's* a completely different use-case, and one we can talk about reasonably, but has nothing to do with the other use-cases based on observable behavior of a conforming implementation.

If the use-case is "we don't expect to ever actually implement `@page { size: ...; }`, at least not for a very long time, but we'd like to be able to have pages that are hacking in landscape-orientation by rendering sideways be oriented 'as intended' when printed to PDF", then the described approach might be reasonable. But then I'd want to be much, much clearer that this is a temporary solution that has better technical solutions that just haven't been implemented yet, and would claim a scarier and less generic name for the property as a result. `orientation` is definitely not an appropriate name for this sort of "feature B to deal with the fact that feature A isn't implemented" feature.

**Comment by chrishtr** [comment]
> As far as I can tell from your use-case description and described behavior, it's the exact opposite. From your description of the use-case, it's for when you're rendering to a PDF or other page-based _screen_ media. In the case you described, all the pages are `size: portrait` already, and thus suitable for immediately printing with no additional work; you use `orientation` to get the supposed-to-be-landscape pages to be shown in landscape _on screen_.

When I said "for printed output", I mean that the printing use case. PDF generation is a kind of printing, it's not "view on the web".

> My suggested inversion was the "just for printed output" version, where screen/PDF display works correctly by default (by using `size` properly), and then you use `margin-orientation` to rotate how the margins are laid out so they're in a desirable spot when printing.

Browsers don't have any distinction between PDF and non-PDF printed output, do they? PDF output is an implementation detail of the user agent outside of the control of the web developer.

> If the use-case is ...

This is another motivation, yes. However, I don't agree with the notion that this is a hack. The other proposed solutions, as far as I can tell, imply a very large amount of implementation work (even more than just implementing full fragmentation, because there is the proposed additional feature of rotating headers and footers differently than the page layout). Therefore this solution is has positive design points for the (valid, non-hack) use-cases stated.

**Comment by faceless2** [comment]
> If the page is landscape then headers will also be landscape, right? Which is not desired.

I think that's false if the page is being displayed as landscape on screen - the pages are oriented to match their "most important direction", all the text is up the right way, no-one has to turn their head sideways to read anything.

But I think it's true if the page is being printed and bound into a book where all the pages are going to wind up as portrait regardless of their original orientation. In that case, yes, as I mentioned before I can see this being a use-case for some sort of parameter. But again, I would personally lean towards the concept of "rotating the margin content sideways" rather than "rotating the page content sideways".

**Comment by chrishtr** [comment]
> I think that's false (i.e. it is desired) if the page is being displayed as landscape on screen - the pages are oriented to match their "most important direction", all the text is up the right way, no-one has to turn their head sideways to read anything.

I see your point that it's more readable on a screen to see everything in the same orientation. But the use case here is a PDF preview of printing, which is outside the control of the web platform, and therefore requires some parameter to help it out. Right?

**Comment by faceless2** [comment]
Sure - you've convinced me that some sort of parameter indicating a page is to be oriented contrary to it's "most important direction" is probably useful. I'm imagining the printed material, you're imagining the screen preview of the same, but I think we're finally on the same page (badum tish).

**Comment by chrishtr** [comment]
> I just want a different name to emphasize that fact! `orientation` is too prime of a name to use for a "we're doing X because we haven't implemented Y yet" feature.

However, there are other legitimate use-cases that will remain even after such features appear in the future (as mentioned earlier, and @faceless2 agreed are legitimate. 

How about `rotate: none | left | right`?

**Comment by tabatkins** [comment]
> However, there are other legitimate use-cases that will remain even after such features appear in the future

Yup, I also agree that having the ability to dictate the relative rotation of the page content vs the margins is worthwhile in the future.  My ideal to achieve the use-case here is:

```css
table.big {
 page: big-table;
}
@page big-table {
 size: landscape;
 mismatched-orientation: rotate-out;
}
```

...which would be normal on screen, would show a landscape page in PDF (with @top on either the left or right), and would be printed on portrait paper such that the @top margin is on top, and the table is rotated in the "out" direction (right on right pages, left on left pages).

> How about `rotate: none | left | right`?

Maybe? Given how targeted this is at the PDF use-case, maybe `pdf-rotate`?

**Comment by faceless2** [comment]
Both `orientation` and `rotate` seem - as Tab puts it - "too prime a name" for what is, in the long term, going to be a bit of an edge-case property.

Not quite sure about `pdf-rotate` either - if someone is targetting, I don't know, PostScript or TIFF, they might still want to do the same. I think if you want values of "left" or "right" or "90deg" or "-90deg" then has to be something-rotation - if you want values of "portrait" or "landscape" then it should be something-orientation.

* Putting lots of pages together to prepare for print is called "assembly" - how about "assembly-rotation" or "assembly-rotate"?
* "print-rotate"? Easily understood this would apply to print-preview as well.
* "page-rotate"? Getting a bit close to "rotate" perhaps
* "orientation-correction"? No one's going to reach for that one instead of "size", that's for sure.

**Comment by chrishtr** [comment]
Thanks for offering suggestions.

"print-rotate", "page-rotate" and"orientation-correction" all sound reasonable to me!

I don't think it's good to put a particular output format like "pdf" in the name.

**Comment by tabatkins** [comment]
Yeah, I wasn't happy with pdf-rotate either. page-rotate seems good to me; you're literally rotating the page itself, but nothing else changes about the layout or display.

**Comment by fantasai** [comment]
Should we be aligning on the value space for `image-orientation`? https://www.w3.org/TR/css-images-3/#the-image-orientation Or perhaps borrow from `text-orientation` and use the prefix `sideways-` instead of `rotate-`?

Also `page-orientation: none` is a bit nonsensical, if we're going with keywords maybe `upright` to match `text-orientation`?

**Comment by tabatkins** [comment]
Yeah, with "orientation" "none" doesn't make sense. "upright" works for me.

("page-rotate: none" was more reasonable)

**Comment by mstensho** [comment]
Explainer: https://github.com/mstensho/page-orientation/
